android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/home/HomeViewModel.kt

```diff
@@ -0,0 +1,48 @@
+package co.nimblehq.ic.kmm.suv.android.ui.screens.home
+
+import androidx.lifecycle.viewModelScope
+import co.nimblehq.ic.kmm.suv.android.ui.base.BaseViewModel
+import co.nimblehq.ic.kmm.suv.domain.usecase.GetProfileUseCase
+import co.nimblehq.ic.kmm.suv.helper.date.DateFormat
+import co.nimblehq.ic.kmm.suv.helper.date.DateTime
+import co.nimblehq.ic.kmm.suv.helper.date.DateTimeFormatter
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.launch
+
+class HomeViewModel(
+    private val getProfileUseCase: GetProfileUseCase,
+    dateTime: DateTime,
+    dateTimeFormatter: DateTimeFormatter
+    ) : BaseViewModel() {
+
+    private val _currentDate = MutableStateFlow("")
+    val currentDate: StateFlow<String> = _currentDate.asStateFlow()
+
+    private val _avatarUrlString = MutableStateFlow("")
+    val avatarUrlString: StateFlow<String> = _avatarUrlString.asStateFlow()
+
+    init {
+        _currentDate.value = dateTimeFormatter.getFormattedString(
+            dateTime.today(),
+            DateFormat.WeekDayMonthDay
+        ).uppercase()
+    }
+
+    fun loadProfile() {
+        showLoading()
+        viewModelScope.launch {
+            getProfileUseCase()
+                .catch { e ->
+                    hideLoading()
+                    showError(e.message)
+                }
+                .collect {
+                    hideLoading()
+                    _avatarUrlString.value = it.avatarUrl
+                }
+        }
+    }
```

Comment:
`hideLoading` could be only at the end

```suggestion
    fun loadProfile() {
        showLoading()
        viewModelScope.launch {
            getProfileUseCase()
                .catch { e ->
                    showError(e.message)
                }
                .collect {
                    _avatarUrlString.value = it.avatarUrl
                }
        }
        hideLoading()
    }
```

Reply:
@luongvo It should be in the `viewModelScope.launch` scope, right?

```suggestion
    fun loadProfile() {
        showLoading()
        viewModelScope.launch {
            getProfileUseCase()
                .catch { e ->
                    showError(e.message)
                }
                .collect {
                    _avatarUrlString.value = it.avatarUrl
                }
			hideLoading()
        }
    }
```

IMO, if we put it at the end of the function, it will show and disappear directly because  `viewModelScope.launch` is an asynchronous task, right?

![Cursor_and_Surveys_–_HomeScreen_kt__Surveys_android_main_](https://user-images.githubusercontent.com/19943832/201021065-35c2ca22-cc51-435a-92c6-739ca65c2a60.png)

I'm thinking about adding `hideLoading` in [onCompletion](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/on-completion.html) of the flow. What do you think?

Reply:
> IMO, if we put it at the end of the function, it will show and disappear directly because viewModelScope.launch is an asynchronous task, right?

Nope, the code will be executed in the right order from showLoading() > Use case executing > hideLoading(). You can try to put a `delay(10000)` after `showLoading`, run the app and see the result. 🤓 

Actually, we can put `showLoading()` inside of `viewModelScope.launch {`.

```suggestion
    fun loadProfile() {
        viewModelScope.launch {
            showLoading()
            delay(10000) // <- testing
            getProfileUseCase()
                .catch { e ->
                    showError(e.message)
                }
                .collect {
                    _avatarUrlString.value = it.avatarUrl
                }
            hideLoading()
        }
    }
```

or
```

```

Reply:
or

```suggestion
    fun loadProfile() {
        getProfileUseCase()
            .flowOn(Dispatchers.IO)
            .onStart { showLoading() }
            .onCompletion { hideLoading() }
            .catch { e ->
                showError(e.message)
            }
            .onEach {
                _avatarUrlString.value = it.avatarUrl
            }
            .launchIn(viewModelScope)
    }
```

Reply:
> Nope, the code will be executed in the right order from showLoading() > Use case executing > hideLoading(). You can try to put a delay(10000) after showLoading, run the app and see the result. 🤓

@luongvo For `hideLoading()` outside the `viewModelScope.launch` block. I think it will hide loading directly. Could you double-check that again?

I added a delay in `GetProfileUseCase`, the loading indicator will disappear directly while it should wait the result from `GetProfileUseCase` 🤔

<img width="505" alt="Screenshot 2022-11-11 at 17 33 54" src="https://user-images.githubusercontent.com/19943832/201322254-b4faa22d-7637-4e36-850d-0ef272fa7c1e.png">

But if I moved `hideLoading()` into the `viewModelScope.launch` block, it works as expected. Thanks for the suggestion 🙌

Reply:
Fixed in https://github.com/suho/kmm-ic/pull/100/commits/d3a1994b201a9005d1b1e198f16bc8219c3cc98e

Reply:
> @luongvo For hideLoading() outside the viewModelScope.launch block. I think it will hide loading directly. Could you double-check that again?

@suho My initial suggestion does not provide `hideLoading` outside of the viewModelScope.launch block 🤔 Actually I see you put `showLoading` outside so I updated the suggestion to move it into the block 😄